basalt/src/config/mod.rs

```diff
@@ -0,0 +1,106 @@
+mod keybinding;
+#[cfg(test)]
+mod test;
+
+use std::{collections::HashMap, fs::read_to_string, path::PathBuf, str::from_utf8};
+
+use etcetera::{choose_base_strategy, home_dir, BaseStrategy};
+use keybinding::KeyBinding;
+use serde::Deserialize;
+
+use crate::app::Action;
+
+pub(crate) use keybinding::Key;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Config {
+    pub keymap: HashMap<Key, Action>,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        let TomlConfig { key_bindings } = TomlConfig::default();
+
+        Self {
+            keymap: key_bindings
+                .expect("The built-in config.toml should contain key bindings")
+                .into_iter()
+                .map(|KeyBinding { key, command }| (key, command.into()))
+                .collect(),
+        }
+    }
+}
+
+impl From<TomlConfig> for Config {
+    fn from(TomlConfig { key_bindings }: TomlConfig) -> Self {
+        let mut config = Self::default();
+        config.merge_key_bindings(key_bindings);
+
+        config
+    }
+}
+
+impl Config {
+    pub fn build() -> Self {
+        TomlConfig::build().map_or_else(Self::default, Self::from)
+    }
+    fn merge_key_bindings(&mut self, key_bindings: Option<Vec<KeyBinding>>) {
+        if let Some(key_bindings) = key_bindings {
+            key_bindings
+                .into_iter()
+                .for_each(|KeyBinding { key, command }| {
+                    let new_action: Action = command.into();
+                    self.keymap
+                        .entry(key)
+                        .and_modify(|old_action| *old_action = new_action)
+                        .or_insert(new_action);
+                })
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize)]
+struct TomlConfig {
+    key_bindings: Option<Vec<KeyBinding>>,
+}
```

Comment:
Let's use the serde default instead of wrapping this in Option!
```suggestion
struct TomlConfig {
    #[serde(default)]
    key_bindings: Vec<KeyBinding>,
}
```

Reply:
Fine by me.